Automattic/jetpack
# Launchpad: "first_post_published" task links to first post if you have one

Fixes https://github.com/Automattic/wp-calypso/issues/94086

## Proposed changes:


Some flows allow the user to start writing before they reach the Launchpad, if that is the case, users can have a Draft post when they arrive at the Launchpad. As reported in https://github.com/Automattic/wp-calypso/issues/94086, the user can find themselves in an onboarding flow loop: they draft a first post, head back to the launchpad, click "write your first post", draft a new post, head back to the launchpad, etc. etc.
The user should be more successful completing the task if they don't end up in a loop.

This PR adds checks on the Launchpad API and to return the correct task URL:
- Links to either an existing draft or an empty editor
- Links to either `wp-admin` or Calypso editors depending on flow and site options
- Returns different copy depending on whether an existing draft exists.

This PR works in concert with https://github.com/Automattic/wp-calypso/pull/94259 which updates the client logic so that it respects the URL returned by the server. That's why extra logic for `start-writing` and `design-first` flows have been added to the server, because it replicates logic that used to be present in Calypso.

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



No

## Testing instructions:






The same testing instructions as the Calypso PR Automattic/wp-calypso#94259 replicated below for convenience.

Extra things to test for this PR are to using logging or some other mechanism to confirm the caching in the new `wpcom_launchpad_get_latest_draft_id` function works as expected.

<details>
* Add the Jetpack changes to your sandbox
  * Sandbox `public-api`
  * Run `bin/jetpack-downloader test jetpack-mu-wpcom-plugin feat/launchpad_draft_post`
  * See PCYsg-Osp-p2 for more details
* Build Calypso changes locally
* Test `write` flow with a draft post
  * Create new site using `/start`
  * When asked about goals choose **Write & Publish**
  * Continue on flow and choose **Draft your first post: Start writing** button
  * Confirm the suggested writing ideas modal appears when the editor loads
    * This is controlled by the `new_prompt=true` query param
  * Save your post as draft but *do not publish*
  * Preview the draft in a new tab
  * In the new tab click **Next steps**
  * You are taken to launchpad, at this point you'll need to change the page domain back to `calypso.localhost:3000`
  * Confirm the **Continue to write your first post** button links back to the editor
    * It should re-open your existing post
    * It should use the Calypso editor
    * There is no `new_prompt` modal when the editor loads
  * Now test in `wp-admin` mode
    * Jump straight to `/settings/general/{{ site slug }}` and change the admin interface to **classic**
    * Click **My Home** in the sidebar to get redirected to the launchpad (confirm your still looking at `calypso.localhost:3000`)
    * Confirm the **Continue to write your first post** button links back to the editor
      * It should re-open your existing post
      * It should use the`wp-admin` editor
      * There is no `new_prompt` modal when the editor loads
* Test `write` flow with no draft post in Calypso site
  * Create new site using `/start`
  * When asked about goals choose **Write & Publish**
  * Continue on flow and choose **Draft your first post: Start writing** button
  * Confirm the suggested writing ideas modal appears when the editor loads
    * This is controlled by the `new_prompt=true` query param
  * *do not save a draft*, leave the editor with the (W) button
  * You are taken to launchpad, at this point you'll need to change the page domain back to `calypso.localhost:3000`
  * Confirm the **Write your first post** button links back to the editor
    * It opens a blank editor
    * It should use the Calypso editor
    * This is a `new_prompt` modal when the editor loads (they might have accidentally dropped out of the flow earlier so I think it's reasonable to show it to them here)
  * Now test in `wp-admin` mode
    * Jump straight to `/settings/general/{{ site slug }}` and change the admin interface to **classic**
    * Click **My Home** in the sidebar to get redirected to the launchpad (confirm your still looking at `calypso.localhost:3000`)
    * Confirm the **Continue to write your first post** button links back to the editor
      * It should open a blank editor
      * It should use the`wp-admin` editor
      * There is a `new_prompt` modal when the editor loads
* Test `start-writing` or `design-first` flow
  * Create a new site using `/setup/start-writing`
  * You are eventually taken to the launchpad
  * Confirm the **Write your first post** button links to the editor
    * It should open a blank editor
    * It should use the `wp-admin` editor (remember these flows are special cased to use `wp-admin`)
    * There is no `new_prompt` modal
    * An `origin` query param has been added
  * Save and preview a draft post
  * In the new tab click **Next steps**
  * Back in the launchpad (confirm you're on `calypso.localhost:3000`) check the **Continue to write your first post** button
    * It should open the existing draft
    * It should use the `wp-admin` editor (remember these flows are special cased to use `wp-admin`)
    * There is no `new_prompt` modal
    * An `origin` query param has been added
</details>
